### PR TITLE
Ranged weapon rebalance

### PR DESCRIPTION
**Pull request by Firestorm01X2**
Attempt to do something with this:
https://github.com/CleverRaven/Cataclysm-DDA/issues/22753

Main ideas of rebalance - increase engagement ranges, make sniping over 30 tiles avalible for average character and give better progression to player accuracy based on skill. At the same time I've tried to use simple approach without reinventing ranged balance.

Features:
* Now both Marksmanship and  Gun Skill have considerable effect on player accuracy.
* Now Gun Skill (like "Rifles", "SMG") actually important and increases chance to hit not just helps with recoil.
* Significantly increased effective range for competent shooter. In most cases it was doubled.
* Reduced minimum gun dispersion to 1. So now weapon stats have almost no cap on how accurate it can be.
* End game player can snipe over all current "reality bubble" as long as weapon allows it.
* Increased effect of CBM Targeting.
* Max weapon range become important part of weapon stats since now experienced player can potentially reach it.
* Changes covered with unit tests.

Idea initially based on:
https://discourse.cataclysmdda.org/t/experimental-builds-with-improved-gun-accuracy-updated-11-05-2018/15131

Outline:
Look for: https://github.com/CleverRaven/Cataclysm-DDA/issues/21244#issuecomment-310013403
to get definition of character types.

Range table:

Character | >50% good | <10% good
-- | -- | --
Novice pistol | 5 | 15
Novice SMG | 7 | 20
Novice rifle | 9 | 25
Competent pistol | 13 | 35
Competent SMG | 20 | 55
Competent rifle | 30 | 90
God pistol | 20 | 120
God SMG | 30 | 190
God rifle | 60 | 800

I've intended to keep current max weapon ranges. Because:
1) Some weapons already have max range sized to "reality bubble".
2) It helps to keep things balanced against melee weapon.
3) It helps to differentiate gun types.

Notes:
* Implementation balanced closer to "average" (competent shooter) character with average skills. 
* "God" character may have godlike accuracy  limited only with weapon stats. It is intended.
* "God" character have <10% chance to hit extended far beyond reality bubble. It is intended since it is endgame character designed to almost always hit if he/she has time to aim. Weapons range still limited with reality bubble. Also note that reaching this accuracy level now require both Marksmanship and weapon skill.

### CODE REVIEW

**Comment by tyrael93** [comment]
Oh boy

**Comment by kevingranade** [comment]
Overall this is what I've been asking for, thank you.

> Changes affects only player and NPCs. Monsters and robots have original accuracy for balance reasons.

This is the opposite of maintaining balance, afaict this means robots are always going to lose in a shootout with an even average player, that is not how it is intended to work.  If I'm wrong and only very high level characters can outshoot robots for some reason, then it's not that big a deal, though I still don't like that it establishes a seperate accuracy curve for players vs non-players.

> Added new dispersion multiplier that decreased weapon dispersion based on average value between Marksmaship and weapon skill level.

Decreasing gun dispersion with skills is not ok, that invalidates the entire ranged accuracy system.
The way to do that is decreasing gun dispersion and moving that dispersion to players with low skills.  The dispersion from low skill can be based on gun dispersion if you like, that should make it easier to maintain scaling.

**Comment by Firestorm01X2** [comment]
>Decreasing gun dispersion with skills is not ok, that invalidates the entire ranged accuracy system.
>The dispersion from low skill can be based on gun dispersion if you like, that should make it easier to maintain scaling.

Going this way. I helps to differentiate guns even if they have low dispersion.
But I have to reduce encumbrance dispersion  and stats dispersion  this way- they way to high.

**Comment by Firestorm01X2** [comment]
Ok. I've bring approach online. I 've to correct range table because now I can't tweak dispersion globally with multiplier. I tried to replicate  previous results. Competent shooter n result get slightly lower effective range then before.

File: src/itype.h [hunk 1]
```diff
@@ -409,7 +409,7 @@ struct islot_gun : common_ranged_data {
     int reload_noise_volume = 0;
 
     /** Maximum aim achievable using base weapon sights */
-    int sight_dispersion = 120;
+    int sight_dispersion = 75;
```
Comment (John-Candlebury) [inline src/itype.h]: Why is this hidden limiter here?  Cant we  just use the actual gun sight dispersion, this just seem unfriendly to modding.

Why do we assume that all factory-issue iron sights must be bad?

Comment (Firestorm01X2) [reply src/itype.h]: Good question. I think that it simulates inability to player precisely see distant targets in sight. It make sense.

Comment (Firestorm01X2) [reply src/itype.h]: I'll lower it.

Comment (John-Candlebury) [reply src/itype.h]: Then why not add the penalty if the actual shot distance supprasses a given treshold. (Maybe taking player PER into account)

That way, pistols and shotguns will not suffer from this.

Comment (Firestorm01X2) [reply src/itype.h]: I propose to leave it for next PR. Lowering it 35 helped already.

**Comment by Firestorm01X2** [comment]
Looks like it is stable. Engagement ranges increased. Skill progression is exist. Sniping over 30 tiles possible.

**Comment by Firestorm01X2** [comment]
Now it is much better. For example effective range of m16 slowly progress from 10 to 25 with increasing skill levels. If you want more than 30 - attach the scope.

I think now is time to stop. I really like the results and feeling of progression.

**Comment by Firestorm01X2** [comment]
Ready for review.

**Comment by thethunderhawk** [comment]
Exciting! I’ve seen a lot of complaints recently that city centers are nearly impenetrable with wander spawns on. Hopefully this newfound ability to pick off the necromancers and hulks from a distance will curb that a bit.

**Comment by Firestorm01X2** [comment]
@kevingranade 
I have question. I had to increase shotgun dispsersion (+800 dipsersion) to mimic previopus shotguns close range behaviour. I did it in jsons but... There are a lot of mods with new weapons. Can I harcode dispersion increase for shotguns? Because without this change shotguns will be crazy accurate (in most cases they have jsons stats of super short range sniper rifles).

**Comment by Wishbringer** [comment]
Also dispersion would depend on the ammunition type, like slugs fired from a Saiga 12g would be quite accurate, in real life.

**Comment by Firestorm01X2** [comment]
>Also dispersion would depend on the ammunition type, like slugs fired from a Saiga 12g would be quite accurate, in real life.

I understand it. But shotguns in game has pretty low range - [0,6].  Shothungs role was- close range high damage weapon.

I don't want to change it. So I want to squize shotgun effective range to that [0, 6] as it was before. At least for now. 
Ammunition type already should be taken in account. But effect will be pretty unnoticable.

Main problem that before shotguns rely on huge sight dispersion. Now it is pretty low. But shotguns surpsiningly had  stat dispersion twice as lower than in sniper rifles.  Without high sight dispersion  it become problem. So I solve it by increasing dispersion. Results pretty close to previous. But overall accuracy slightly higher.

**Comment by Wishbringer** [comment]
Update: Seems like the shot_slug overrides the dispersion to 100, so would probably be fine as is?
BTW, you're doing an awesome job on this, its very difficult to balance these kind of things.

**Comment by Firestorm01X2** [comment]
>Update: Seems like the shot_slug overrides the dispersion to 100,

I think it adds to it acutally to make accuracy even worse
>"description": "A heavy metal slug used with shotguns to give them the range capabilities of a rifle.  Extremely damaging but rather *inaccurate*.",

**Comment by Wishbringer** [comment]
Depends on whether or not the shotgun barrel is rifled, but adding dispersion to slugs is clearly not correct, especially with the new shotgun dispersion values.

**Comment by Firestorm01X2** [comment]
>Depends on whether or not the shotgun barrel is rifled, but adding even more dispersion to slugs is clearly not correct, especially with the new shotgun dispersion values.

Overall dispersion of gun = gun dispersion + gunmods dispersion + **ammo dispersion**

It works that way. Slugs by design increase range but at the same time decrease dispersion to make target harder to hit. It was done before that way.

>"description": "A heavy metal slug used with shotguns to give them the range capabilities of a rifle.  Extremely damaging but rather inaccurate.",

**Comment by Firestorm01X2** [comment]
I've put shotgun in range of dispersion aftificially via code. This way mods like rifled barrel take effect. Still not perfect but at least it works.

**Comment by boydkr** [comment]
> But then again I can't remember whether the individual pellets are calculated separately or not.

One reasonable approximation of shotgun accuracy / damage would be to leave dispersion fairly low, but model multiple pellets by having damage fall off with range (modeling fewer projectiles hitting the target)

**Comment by Firestorm01X2** [comment]
I guess it's time to say that it is ready.

**Comment by BorkBorkGoesTheCode** [comment]
@Firestorm01X2 instead of tying shotgun dispersion to guns using the shotgun skill, could you use the SHOT flag? Most shot ammo already has the flag.

**Comment by John-Candlebury** [comment]
Why not just add the dispersion to the 00 shot and flechte shells?

There are I think for types of shotgun ammo in the game, so it houldn't be much work, and that way remains 100% modable

File: data/json/items/ammo/shot.json [hunk 2]
```diff
@@ -54,6 +54,7 @@
     "damage": 50,
     "recoil": 2500,
     "loudness": 160,
+	"dispersion": 800,
```
Comment (Wishbringer) [inline data/json/items/ammo/shot.json]: Tab rather than spaces.

File: data/json/items/handloaded_bullets.json [hunk 3]
```diff
@@ -31,6 +31,7 @@
         "ammo_type" : "blunderbuss",
         "damage" : 60,
         "range" : 3,
+		"dispersion" : 800,
```
Comment (Wishbringer) [inline data/json/items/handloaded_bullets.json]: Tab rather than spaces.

File: data/json/items/handloaded_bullets.json [hunk 4]
```diff
@@ -48,6 +49,7 @@
         "damage" : 53,
         "pierce" : 12,
         "range" : 5,
+		"dispersion" : 800,
```
Comment (Wishbringer) [inline data/json/items/handloaded_bullets.json]: Tabs

**Comment by Firestorm01X2** [comment]
>Why not just add the dispersion to the 00 shot and flechtete shells?
There are I think for types of shotgun ammo in the game, so it shouldn't be much work, and that way remains 100% modable

Done. Looks like it is worked out.

**Comment by Firestorm01X2** [comment]
I has to end somewhere. I think it is time to stop tweaking.

Ready.

**Comment by Firestorm01X2** [comment]
Jenkins, Rebuid.

**Comment by Firestorm01X2** [comment]
Jenkins rebuild

**Comment by Firestorm01X2** [comment]
> skill boost for having the targeting CBM doesn't impact the tests and shouldn't be present in this PR

In fact in is part of dispersion calculation. So it has to be taken in account also.

From realisation perspective:
I did it because I wanted to get noticable effect for CBM targeting. Current
`dispersion.add_multiplier( 0.75 );`
now may give much less effect since overall dispersion are lower.

Additonaly at some point I wanted to remove multiplier compeletely  to left only skill boost part (overall player has CBM not hte gun itself). But it make usefeull to install that CBM even if player at max skill already.

From logic perspective:
This CBM by its description do 2 things:
1) Control player hands partially
2) Project some kind of range finder to player sight.

First  can be considered as skill boost part.
Second - as overall dispersion reduction by multiplier.

**Comment by Firestorm01X2** [comment]
Thank you! 
Finally rebalance is here.

**Comment by BorkBorkGoesTheCode** [comment]
I'm going to update the shotgun shells, unless you want to.

**Comment by Firestorm01X2** [comment]
>I'm going to update the shotgun shells, unless you want to.

I've added dispersion already.

But go ahead if you want to change something.

**Comment by BorkBorkGoesTheCode** [comment]
@narc0tiq the build system appears to be repeating itself and the console display has been disabled.

**Comment by BorkBorkGoesTheCode** [comment]
I was talking about the 20x66s .22 ratshots and .410s: they're still have the old dispersion values.

**Comment by Firestorm01X2** [comment]
>I was talking about the 20x66s .22 ratshots and .410s: they're still have the old dispersion values.

Ow I don't touch them. Yes do it.

As I said I've added +800 dispersion.

**Comment by narc0tiq** [comment]
@BorkBorkGoesTheCode 
> The WIndows x64 build is marked dirty.

I think that's from the file `src/crafting.cpp` that managed to get committed with CRLF line endings (somehow ignoring the `.gitattributes`, which wants to put it back to normal).

I guess I should PR a fix to that, else everything's going to be marked `-dirty` forever.

**Comment by BorkBorkGoesTheCode** [comment]
Increased shot dispersion by 800 here: https://github.com/CleverRaven/Cataclysm-DDA/pull/23747

**Comment by BorkBorkGoesTheCode** [comment]
@Firestorm01X2 why were LIFT requirements changed here https://github.com/CleverRaven/Cataclysm-DDA/pull/23720/commits/930de0a0b9ba3a52a667f93a939b2732f40eb88b?

**Comment by Firestorm01X2** [comment]
>@Firestorm01X2 why were LIFT requirements changed here 930de0a?

>It is reverted in e5b8338.

@BorkBorkGoesTheCode 

Indeed. This PR branch based of my old rebalance fork branch. I did there not only accuracy changes. But it was reverterd, yes.

**Comment by vipches** [comment]
Hi, Firestorm, I actually have one thing to ask, what about the monster who can shooting too? So far as we know, those monster, like turrets, I don't know why, but they seems can actually shoot any target without miss a single bullet, so does the NPC too, and the robots like W11B10 or something like that, can almost never miss the shot on the target, even in a far distance, I remember when I wear a full power armor going to the base, and those robot just shoot me directly without miss, I try to shoot them in the same distance that they can shoot me, and I found out I can't actually catch up them, they always doing shooting one step before me, and anytime I taking hit(will no damage) I reduce my accuracy, and has to start all over, but due they are faster aiming the target than me, I can't doing anything to shoot back, cause I'm keep taking hit, the same thing happen on the turrets too, and something the NPC can took the shoot first too, and they are keep going and going, make you unable to doing anything about it, So, I want to know, if there is anything that can doing about this situation? al least make those shooter have a chance to miss the target?

**Comment by Firestorm01X2** [comment]
@vipches 
Yes You are right.
About robots:
1) Robots has high skills.
2) Robots has no encumbrance
3) Robots don't need to aim.
4) Most shooting robots are small so it is twice as harder to hit them.

Robots was accurate before. But now they are crazy accurate. Just like you in most cases.

And looks like it is just like planned. Look at Kevin's response:
https://github.com/CleverRaven/Cataclysm-DDA/pull/23720#issuecomment-388591159
>This is the opposite of maintaining balance, afaict this means robots are always going to lose in a shootout with an even average player, that is not how it is intended to work. If I'm wrong and only very high level characters can outshoot robots for some reason, then it's not that big a deal, though I still don't like that it establishes a seperate accuracy curve for players vs non-players.

Initially I've planned to reduce accuracy bonus for non NPC. But idea was declined.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/request-to-developers/22170/1

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/apologies-to-community/22942/1

**Pull request merged**